builder/builder.go

```diff
@@ -29,27 +31,100 @@ type Builder struct {
 	privKey    *ecdsa.PrivateKey
 
 	bc       *blockchain.Blockchain
+	pool     *mempool.Pool
 	vm       vm.VM
 	newHeads *feed.Feed[*core.Header]
 	log      utils.Logger
+
+	pendingLock  stdsync.Mutex
+	pendingBlock blockchain.Pending
+	headState    core.StateReader
+	headCloser   blockchain.StateCloser
 }
 
-func New(privKey *ecdsa.PrivateKey, ownAddr *felt.Felt, bc *blockchain.Blockchain, builderVM vm.VM, log utils.Logger) *Builder {
+func New(privKey *ecdsa.PrivateKey, ownAddr *felt.Felt, bc *blockchain.Blockchain, pool *mempool.Pool, builderVM vm.VM, log utils.Logger) *Builder {
 	return &Builder{
 		ownAddress: *ownAddr,
 		privKey:    privKey,
 
 		bc:       bc,
+		pool:     pool,
 		vm:       builderVM,
 		newHeads: feed.New[*core.Header](),
+		log:      log,
 	}
 }
 
 func (b *Builder) Run(ctx context.Context) error {
-	<-ctx.Done()
+	if err := b.initPendingBlock(); err != nil {
+		return err
+	}
+	defer b.clearPending()
+
+	go b.listenPool(ctx)
+
+	const blockTime = 5 * time.Second
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(blockTime):
+			if err := b.Finalise(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (b *Builder) initPendingBlock() error {
+	if b.pendingBlock.Block != nil {
+		return nil
+	}
+
+	bcPending, err := b.bc.Pending()
+	if err != nil {
+		return err
+	}
+
+	b.pendingBlock, err = utils.Clone[blockchain.Pending](bcPending)
+	if err != nil {
+		return err
+	}
+	b.pendingBlock.Block.SequencerAddress = &b.ownAddress
+
+	b.headState, b.headCloser, err = b.bc.HeadState()
+	return err
+}
+
+func (b *Builder) clearPending() error {
+	b.pendingBlock = blockchain.Pending{}
+	if b.headState != nil {
+		if err := b.headCloser(); err != nil {
+			return err
+		}
+		b.headState = nil
+		b.headCloser = nil
+	}
 	return nil
 }
 
+// Finalise the pending block and initialise the next one
+func (b *Builder) Finalise() error {
+	b.pendingLock.Lock()
+	defer b.pendingLock.Unlock()
+
+	if err := b.bc.Finalise(&b.pendingBlock, b.Sign); err != nil {
+		return err
+	}
+	b.log.Infow("Finalised block", "number", b.pendingBlock.Block.Number, "hash",
+		b.pendingBlock.Block.Hash.ShortString(), "state", b.pendingBlock.Block.GlobalStateRoot.ShortString())
+
+	if err := b.clearPending(); err != nil {
+		return err
+	}
+	return b.initPendingBlock()
```

Comment:
nit: do you think we should just return nil, and call clearPending() / initPendingBlock() outside of Finalise()?

Reply:
We could and actually should in the long run, for better and more granular error handling. But we can keep it the way it is for simplicity for now.